### ISSUE TITLE
Modded Enemy Weights Don't Function Properly

### ISSUE THREAD

**Issue by Sudux17** [BananaPuncher714/LethalQuantities#108]
I'm having issues with modded enemies spawning.

I am working on modpack with a WHOLE bunch of moons, interiors, and monsters. I wanted to text out some of the new monsters, so I used LethalQuantities to set the weight of all interior monsters for Experimentation to 0, as well as setting that moon's power level to 10. I then set the weight for the modded monsters I wanted to test first (Baldi and the HauntedHarpist, if that matters) to 20 each. Everything seemed to be working properly, so I went in, set theirs weights back to 0, and set the weight of three more modded creatures (Herobrine, the Immortal Snail, and an entity called the "knight" that I didn't recognize).
However, I could not get these enemies to spawn. For the most part, no matter what I did, the only monsters that would spawn on Experimentation were Baldi and the Haunted Harpist. There were two fringe cases where other entities spawned, but only in 1 case was it an entity who's current spawn weight was 20 (Herobrine), and the other case it was a completely seperate modded entity (the SCP Shy Guy) who's spawn weight had never been set to 20. No vanilla creatures ever spawned. Additionally, to the best of my memory, the total power of the monsters never went above 4, which is the moon's base power, instead of spawning enemies up to the new power level of 10. This held true even in the two cases when other monsters spawned.
I tried a lot to try to get it to work. I restarted Lethal Company, I restarted r2ModManager, disabled and re-enabled LethalQuantities, I deleted the save file, I tried a different save file, and I even deleted the config files, regenerated them, and set all the weights to 0, this time only setting the second group's weights to 20. Nothing makes a difference.
I really hope I'm not just being stupid and missing something obvious.

Playing on Lethal Company v68
r2ModManager v3.1.54
LethalQuantities v1.2.9

**Comment by BananaPuncher714** [comment]
Do you have any other mods that might change spawn weights? They could interfere with the allowed enemies. Herobrine may have a non-vanilla spawn mechanic which would enable it to spawn regardless of any settings, and the shy guy is known to bypass LQ by changing the weights after LQ loads. Unfortunately, there are a handful of enemies that do not seem to work with LQ.

I think the immortal snail and knight should have worked though, I've used them in the past. Can you try modifying only vanilla enemies and reporting back? If it works fine with vanilla enemies then it most likely means the mod has done some tinkering after LQ. Also, to prevent enemies from spawning it is better to remove them from the list than set their weight to 0.

In addition, to make sure certain enemies spawn, try setting their weight to some large number(1000 or something), and change their spawn curve to a single pair of numbers (0, 1).

**Comment by Sudux17** [comment]
After much more testing, I've concluded that it must be a glitch on my end. I've now somehow managed to get my game to a state where I can't spawn ANY monsters despite having LethalQuantities uninstalled, all the config files deleted, and even LethalLevelLoader and all the custom monster mods disabled. The only way I can spawn monsters is by launching without any mods.
If it's any help, I did find that at least Baldi definitely spawns independantly of LethalQuantities before I got too far down the rabbit hole.

**Comment by Sudux17** [comment]
Coming back to this several months later.

I think I've discovered the issue, and its much simpler than I thought. The game takes whatever weight values you have set the first time you run the LethalQuantities and keeps using them even after the values are changed. The only way I've been able to actually change the spawn weights is by uninstalling LethalQuantities and deleting all its config files, then reinstalling LethalQuantities and regenerating the config files (which means I have to restart the game twice before I can test again, once to generate the global config file and then once more to generate Experimentation's config files).

For instance, I set Experimentation to spawn only Blobs and Centipedes at equal weights (1), and only Centipedes spawned. Afterwards I set it to only spawn Blobs and a modded enemy (Shy Guy from the Scocophobia mod) at equal weights (1). Only Centipedes spawned. I did a full reset and tried again with only Blobs and the Shy Guy at equal weights (3). Only Blobs spawned. I then set it to spawn only Centipedes and Shy Guy at equal weights, but got only Blobs (Shy Guy seems to use its own mod's spawn weight over that of LethalQuantities).

Another weird little bug (I don't know whether or not this pertains to LethalQuantities) but Blobs only spawn when their weight was set to at least 2. This did not affect Centipedes, who spawned normally when their weight was set to 1, and I haven't test any other monsters. This is the only relevant text I could find in the command prompt:

[Info   : Unity Log] enemy rush index is -1; current index 0
[Info   : Unity Log] Probability: 0; enemy type: Blob
[Info   : Unity Log] enemy rush index is -1; current index 1
[Info   : Unity Log] Probability: 0; enemy type: Shy guy

P.S. All of this was with the LethalLevelLoader's content configuration setting for Experimentation set to false. I do have LethalLevelLoader's content configuration setting for all vanilla interiors and a bunch of custom interiors set to true, if that matters.

**Comment by BananaPuncher714** [comment]
There shouldn't really be any persistent changes whenever LethalQuantities starts, so I'm not sure what this is caused by. I'll try to narrow down the cause, but I'm not sure this is a LethalQuantities issue. I'll also look into the blob weights, but it is definitely strange, since a weight of 1 should still work regardless of the enemy.

**Comment by Sudux17** [comment]
I think I figured out why Blobs don't spawn when their spawn weight is set to 1 but Centipedes do; It's because the Blob has a spawn chance curve which very slightly lowers its spawn chance throughout the day, while Centipedes have a spawn chance curve which is just a straight line. I'm pretty sure the algorithm rounds down, because at midday this multiplier is 0.99 and by the end of the day it only drops down to 0.95. But with the initial spawn chance set to 1, it will always round down to 0 after multiplying by the spawn chance curve.

Blob spawn chance curve:
0:1, 0.5870568:0.9988297, 1:0.9582062
Centipede spawn chance curve:
0:1, 1:1

**Comment by BananaPuncher714** [comment]
Yes, that would definitely cause a problem. You can try multiplying all the weights by 100 which should fix that particular issue.
